internal/metrics/export/otel.go

```diff
@@ -0,0 +1,242 @@
+// LICENSE file in the root directory of this source tree.
+package export
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/config"
+	"github.com/nginx/agent/v3/internal/model"
+	"github.com/sagikazarmark/slog-shim"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/sdk/instrumentation"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	agentScope = "github.com/agent/v3"
+	// tenantID is static for now.
+	tenantID         = "7332d596-d2e6-4d1e-9e75-70f91ef9bd0e"
+	serviceNamespace = "nginx"
+)
+
+type OTelExporter struct {
+	conf           config.Config
+	intExp         *otlpmetricgrpc.Exporter
+	convert        model.Converter[metricdata.Metrics]
+	mut            *sync.Mutex
+	sink           chan model.DataEntry
+	buffer         []metricdata.Metrics
+	bufferLen      int
+	retryCount     int
+	exportInterval time.Duration
+	res            *resource.Resource
+	scope          instrumentation.Scope
+}
+
+// NewGRPCExporter returns a OTel export that transmits via gRPC.
+func NewGRPCExporter(ctx context.Context, conf config.GRPC) (*otlpmetricgrpc.Exporter, error) {
+	ctx, cancel := context.WithTimeout(ctx, conf.ConnTimeout)
+	defer cancel()
+	// Exponential back-off strategy.
+	backoffConf := backoff.DefaultConfig
+	// You can also change the base delay, multiplier, and jitter here.
+	backoffConf.MaxDelay = conf.BackoffDelay
+
+	conn, err := grpc.DialContext(
+		ctx,
+		conf.Target,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoffConf,
+			// Connection timeout.
+			MinConnectTimeout: conf.MinConnTimeout,
+		}),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create grpc connection: %w", err)
+	}
+
+	return otlpmetricgrpc.New(
+		ctx, otlpmetricgrpc.WithGRPCConn(conn), otlpmetricgrpc.WithTimeout(conf.ConnTimeout),
+	)
+}
+
+func NewOTelExporter(ctx context.Context, agentConf config.Config, serviceName string,
+	c model.Converter[metricdata.Metrics],
+) (*OTelExporter, error) {
+	exp, err := initGPRCExporter(ctx, agentConf.Metrics)
+	if err != nil {
+		return nil, fmt.Errorf("failed to instantiate gRPC exporter for OTel collector: %w", err)
+	}
+
+	scope := instrumentation.Scope{
+		Name:    agentScope,
+		Version: agentConf.Version,
+	}
+
+	res, err := resource.New(ctx,
+		// Keep the default detectors
+		resource.WithTelemetrySDK(),
+		// Add your own custom attributes to identify your application
+		resource.WithAttributes(
+			semconv.ServiceNamespaceKey.String(serviceNamespace),
+			semconv.ServiceNameKey.String(serviceName),
+			semconv.ServiceInstanceIDKey.String(tenantID),
```

Comment:
Discussed this at the office, decided to generate a unique UID for now so that each Agent at least has a unique identifier. The UUID will not survive an Agent restart, so we will need to figure out a way to make it persistent in the future.